### PR TITLE
[REVIEW] Adding new variable to support switching GLIBCXX11 ABI

### PR DESCRIPTION
**Pull request by mt-jones**
Adding a new CMake flag to our project to quickly turn on/off the GLIBCXX11 ABI:

```
###################################################################################################
# - compiler options ------------------------------------------------------------------------------
option(CMAKE_CXX11_ABI "Enable the GLIBCXX11 ABI" ON)
...
```

The default is `ON`, and usage looks like this

```bash
$ cmake .. -DCMAKE_CXX11_ABI=OFF ...
```

### CODE REVIEW

File: cpp/cmake/Modules/ConfigureArrow.cmake [hunk 1]
```diff
@@ -1,23 +1,45 @@
...
+                         " -DARROW_BUILD_TESTS=OFF"
+                         " -DARROW_TEST_MEMCHECK=OFF"
+                         " -DARROW_BUILD_BENCHMARKS=OFF"
+                         " -DARROW_IPC=ON"
+                         " -DARROW_COMPUTE=OFF"
+                         " -DARROW_GPU=OFF"
+                         " -DARROW_JEMALLOC=OFF"
+                         " -DARROW_BOOST_VENDORED=OFF"
+                         " -DARROW_PYTHON=OFF"
+                         " -DARROW_TENSORFLOW=ON") # enable old ABI for C/C++
```
Comment (kkraus14) [inline cpp/cmake/Modules/ConfigureArrow.cmake]: May want to see if we can swap only the string in the future for code clarity, but this is fine for now.

**Comment by mt-jones** [comment]
I expect CI to fail until it's using libraries, etc. which use the new ABI. Otherwise, there will be many missing symbol errors.

File: cpp/cmake/Modules/ConfigureArrow.cmake (hunk 1 above)
Comment (mt-jones) [reply cpp/cmake/Modules/ConfigureArrow.cmake]: `ARROW_CMAKE_ARGS` is a list of strings. I could remove the `" -DARROW_TENSORFLOW={ON/OFF}"` and use `list(APPEND ARROW_CMAKE_ARGS " -DARROW_TENSORFLOW={ON/OFF}"` as a function of the value `CMAKE_CXX11_ABI`.

**Comment by kkraus14** [comment]
rerun tests

**Comment by kkraus14** [comment]
rerun tests

**Comment by mike-wendt** [comment]
rerun tests

**Comment by mike-wendt** [comment]
rerun tests

**Comment by mike-wendt** [comment]
NVStrings issue is fixed, both conda builds and from source build for GPU testing are showing this error:
```
../libcudf.so: undefined reference to 'arrow::ipc::internal::json::JsonWriter::Finish(std::string*)'
```

**Comment by kkraus14** [comment]
rerun tests

**Comment by mike-wendt** [comment]
@mt-jones @harrism @kkraus14 can you please help me understand where we are with this PR to get this out the door

* Am I correct to assume that we should focus on `gpuCI/cudf-prb` as our test base case as it does from scratch builds?
* Do we need gcc7 for this to work or is gcc5 ok?
* If we are building from scratch why do I see the `../libcudf.so: undefined reference to 'arrow::ipc::internal::json::JsonWriter::Finish(std::string*)'` error? Is this due to a wrong gcc version? Or incorrect call when building?

File: cpp/cmake/Modules/ConfigureArrow.cmake (hunk 1 above)
Comment (harrism) [reply cpp/cmake/Modules/ConfigureArrow.cmake]: @mt-jones please follow up on this.

**Comment by kkraus14** [comment]
> * Am I correct to assume that we should focus on `gpuCI/cudf-prb` as our test base case as it does from scratch builds?

Yes we should focus on that as ABI related errors could show during tests via pyarrow.

> * Do we need gcc7 for this to work or is gcc5 ok?

gcc5 is ok and preferred, we haven't really tested gcc7 yet.

> * If we are building from scratch why do I see the `../libcudf.so: undefined reference to 'arrow::ipc::internal::json::JsonWriter::Finish(std::string*)'` error?
>   * Is this due to a wrong gcc version? Or incorrect call when building (like a missing flag)?

I'd defer to @mt-jones but I believe this is because when installing pyarrow from conda it's pulling arrow-cpp from conda still built on the old ABI and that library is being found so we get an ABI mismatch. Sounds like a cmake related issue that @mt-jones may need to investigate.

> * Where have you successfully built this? What was the environment?
>   * The concern comes that if we push this as a hotfix the builders below are the same as the ones that would push a public conda build. So if it fails here then it will fail merged into master.

@mt-jones should answer this, and we should make it work as expected in gpuCI and jarvis before merging this.

**Comment by pearu** [comment]
FYI, conda-forge has recently completed [migration to the new CXX11 ABI](https://conda-forge.org/status/). It means that in near future many related conflicts of using conda and system libraries together will be resolved, or at least, the conflicts will change their nature (hopefully to nothing).

**Comment by harrism** [comment]
@pearu that is the origin of this PR.

File: cpp/cmake/Modules/ConfigureArrow.cmake [hunk 2]
```diff
@@ -38,14 +61,14 @@ endif(ARROW_CONFIG)
 unset(PARALLEL_BUILD)            
 if($ENV{TRAVIS})
     if(NOT DEFINED ENV{CMAKE_BUILD_PARALLEL_LEVEL})
-        message(STATUS "Disabling Parallel CMake build on Travis")
+        message(STATUS "ARROW BUILD: Disabling Parallel CMake build on Travis")
```
[the remark this answers is not in the document]
Comment (mt-jones) [reply cpp/cmake/Modules/ConfigureArrow.cmake]: Sure.

[the remark this answers is not in the document]
Comment (mt-jones) [reply cpp/cmake/Modules/ConfigureArrow.cmake]: This has been done. Merged into PR https://github.com/rapidsai/cudf/pull/668

File: cpp/cmake/Modules/ConfigureArrow.cmake (hunk 1 above)
Comment (mt-jones) [reply cpp/cmake/Modules/ConfigureArrow.cmake]: This has been done. Merged into PR https://github.com/rapidsai/cudf/pull/668

**Comment by mt-jones** [comment]
Closing this pull request as its core changes are merged into PR https://github.com/rapidsai/cudf/pull/668

**Pull request closed without merging**